File: solution/java/week02-object-calisthenics/exercise2/src/main/java/BookStore.java
import java.util.Optional;

public class BookStore {
    BookInventory inventory = new BookInventory();

    public void addBook(String title, String author, int copies) {
        if (copies <= 0)
            return;

        getBookBy(title, author)
            .ifPresentOrElse(
                b -> b.addCopies(copies),
                () -> Book.tryCreateBook(title, author, copies)
                    .ifPresent(inventory::add));
    }

    public void sellBook(String title, String author, int copies) {
        if (copies <= 0)
            return;

        getBookBy(title, author)
            .ifPresent(book -> {
                book.removeCopies(copies);
                inventory.removeBookIfNoMoreCopies(book);
            }
        );
    }

    protected Optional<Book> getBookBy(String title, String author) {
        return inventory.stream()
                .filter(b ->
                    b.getTitle().equals(title)
                    && b.getAuthor().equals(author))
                .findFirst();
    }
}

File: solution/ts/week02-object-calisthenics/exercise2/src/bookStore.ts
import {fold, Option} from 'fp-ts/lib/Option';
import {Book} from './book';
import {BookInventory} from './bookInventory';

export class BookStore {
    private readonly inventory: BookInventory = new BookInventory();

    addBook(title: string | null, author: string | null, copies: number): void {
        if (copies <= 0 || !title || !author) return;

        fold(
            () => {
                const createdBook = Book.createBook(title, author, copies);
                fold(
                    () => {
                    },
                    (book: Book) => this.inventory.addBook(book)
                )(createdBook);
            },
            (book: Book) => book.addCopies(copies)
        )(this.inventory.getBook(title, author));
    }

    sellBook(title: string, author: string, copies: number): void {
        if (copies <= 0) return;

        const bookOption = this.inventory.getBook(title, author);

        fold(
            () => {
            },
            (book: Book) => {
                book.removeCopies(copies);
                this.inventory.removeBookIfNoMoreCopies(book);
            }
        )(bookOption);
    }

    getBookBy = (title: string | null, author: string | null): Option<Book> => this.inventory.getBook(title, author);
}